src/multirust/command.rs

```diff
@@ -0,0 +1,109 @@
+use std;
+use std::env;
+use std::ffi::OsStr;
+use std::io::{self, Write};
+use std::path::PathBuf;
+use std::process::{self, Command, Output};
+use std::time::Instant;
+
+use Cfg;
+use errors::*;
+use multirust_utils;
+use telemetry;
+use telemetry::TelemetryEvent;
+
+
+pub fn run_command_for_dir<S: AsRef<OsStr>>(cmd: Command,
+                                            args: &[S],
+                                            cfg: &Cfg) -> Result<()> {
+    let arg0 = env::args().next().map(|a| PathBuf::from(a));
+    let arg0 = arg0.as_ref()
+        .and_then(|a| a.file_name())
+        .and_then(|a| a.to_str());
+    let arg0 = try!(arg0.ok_or(Error::NoExeName));
+    if (arg0 == "rustc" || arg0 == "rustc.exe") && cfg.telemetry_enabled() {
+        return telemetry_rustc(cmd, &args, &cfg);
+    }
+    
+    run_command_for_dir_without_telemetry(cmd, &args)
+}
+
+fn telemetry_rustc<S: AsRef<OsStr>>(cmd: Command, args: &[S], cfg: &Cfg) -> Result<()> {
+    let now = Instant::now();
+
+    let output = bare_run_command_for_dir(cmd, &args);
+
+    let duration = now.elapsed();
+
+    let ms = (duration.as_secs() as u64 * 1000) + (duration.subsec_nanos() as u64 / 1000 / 1000);
+
+    match output {
+        Ok(out) => {
+            let exit_code = out.status.code().unwrap_or(1);
+
+            let errors = match out.status.success() {
+                true => None,
+                _ => Some(String::from_utf8_lossy(&out.stderr).to_string()),
+            };
+
+            let _ = io::stdout().write(&out.stdout);
+            let _ = io::stdout().flush();
+            let _ = io::stderr().write(&out.stderr);
+            let _ = io::stderr().flush();
```

Comment:
This simple method of capturing output is not going to be sufficient because it delays output until rustc has completed. Instead we're going to have to set up the `Command` to inherit stdout and pipe stderr to our own handle. Then we can loop over reads of our stderr pipe, parse out the error numbers, then print the buffer to the real stderr. This will have the additional benefit of not adding extra overhead to stdout since we're only capturing stderr.